File: src/main/scala/fortress/transformers/SimplifyTransformer.java
package fortress.transformers;

import fortress.tfol.*;

/** 
* Attempts to simplify the formulas in a theory, replacing them with equivalent formulas.
*/
public class SimplifyTransformer implements TheoryTransformer {
    @Override
    public Theory apply(Theory theory) {
        Signature sig = theory.getSignature();
        Theory result = Theory.mkTheoryWithSignature(sig);
        
        for(Term axiom : theory.getAxioms()) {
            Term axiomPrime = axiom.simplify();
            result = result.withAxiom(axiomPrime);
        }
        
        return result;
    }
    
    @Override
    public String getName() {
        return "Simplify Transformer";
    }
}

File: src/main/scala/fortress/transformers/SimplifyTransformer.scala
package fortress.transformers

import fortress.msfol._
import fortress.operations.TermOps._
import fortress.operations.TheoryOps._

/** Applies a simplification to the formulas in a theory, replacing them with equivalent formulas.
  * All other aspects of the theory remain unchanged. */
class SimplifyTransformer extends TheoryTransformer {
    
    override def apply(theory: Theory): Theory =  {
        var learnedLiterals : Map[Term,LeafTerm] = Map.empty
        val newAxioms = theory.axioms.map(axiom => {
            val newAxiom = axiom.simplify(learnedLiterals)
            newAxiom match {
                case Not(Eq(t1, t2)) => learnedLiterals = learnedLiterals + (Eq(t1, t2) -> Bottom)
                case Not(App(fname, args)) => learnedLiterals = learnedLiterals + (App(fname, args) -> Bottom)
                case App(_, _) | Eq(_, _) => learnedLiterals = learnedLiterals + (newAxiom -> Top)
                case _ =>
            }
            newAxiom
        }).filter(_ != Top)
        Theory(theory.signature, newAxioms)
    }
    
    override def name: String = "Simplify Transformer"
}
